test/rule_definition_tools.jl

```diff
@@ -0,0 +1,64 @@
+@testset "rule_definition_tools.jl" begin
+    @testset "@non_differentiable" begin
+        @testset "two input one output function" begin
+            nondiff_2_1(x, y) = fill(7.5, 100)[x + y]
+            @non_differentiable nondiff_2_1(::Any, ::Any)
+            @test frule((Zero(), 1.2, 2.3), nondiff_2_1, 3, 2) == (7.5, DoesNotExist())
+            res, pullback = rrule(nondiff_2_1, 3, 2)
+            @test res == 7.5
+            @test pullback(4.5) == (NO_FIELDS, DoesNotExist(), DoesNotExist())
+        end
+
+        @testset "one input, 2-tuple output function" begin
+            nondiff_1_2(x) = (5.0, 3.0)
+            @non_differentiable nondiff_1_2(::Any)
+            @test frule((Zero(), 1.2), nondiff_1_2, 3.1) == ((5.0, 3.0), DoesNotExist())
+            res, pullback = rrule(nondiff_1_2, 3.1)
+            @test res == (5.0, 3.0)
+            @test isequal(
+                pullback(Composite{Tuple{Float64, Float64}}(1.2, 3.2)),
+                (NO_FIELDS, DoesNotExist()),
+            )
+        end
+
+        @testset "constrained signature" begin
+            nonembed_identity(x) = x
+            @non_differentiable nonembed_identity(::Integer)
+
+            @test frule((Zero(), 1.2), nonembed_identity, 2) == (2, DoesNotExist())
+            @test frule((Zero(), 1.2), nonembed_identity, 2.0) == nothing
+
+            res, pullback = rrule(nonembed_identity, 2)
+            @test res == 2
+            @test pullback(1.2) == (NO_FIELDS, DoesNotExist())
+
+            @test rrule(nonembed_identity, 2.0) == nothing
+        end
+
+        @testset "Pointy UnionAll constraints" begin
+            pointy_identity(x) = x
+            @non_differentiable pointy_identity(::Vector{<:AbstractString})
+
+            @test frule((Zero(), 1.2), pointy_identity, ["2"]) == (["2"], DoesNotExist())
+            @test frule((Zero(), 1.2), pointy_identity, 2.0) == nothing
+
+            res, pullback = rrule(pointy_identity, ["2"])
+            @test res == ["2"]
+            @test pullback(1.2) == (NO_FIELDS, DoesNotExist())
+
+            @test rrule(pointy_identity, 2.0) == nothing
+        end
+        
+        @testset "Not supported (Yet)" begin
+            # Varargs are not supported
+            @test_throws Exception @macroexpand(@non_differentiable vararg1(xs...))|
```

Comment:
typo?
```suggestion
            @test_throws Exception @macroexpand(@non_differentiable vararg1(xs...))
```

Reply:
Just testing for `Exception` is always a bit brittle, because this could also fail for completely unrelated reasons. (Unfortunately the `var` string macro is only available in 1.4, so this is a bit ugly.)
```suggestion
            at_non_differentiable(ex) = getfield(ChainRulesCore, Symbol("@non_differentiable"))(LineNumberNode(@__LINE__, @__FILE__), Main, ex)
            @test_throws ErrorException at_non_differentiable(:(vararg1(xs...)))
```

Reply:
I have added a helper macro just for testing this.